File: cpp-client/deephaven/tests/ticking_test.cc
/*
 * Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
 */
#include <condition_variable>
#include <iostream>
#include <mutex>
#include "tests/third_party/catch.hpp"
#include "tests/test_util.h"
#include "deephaven/client/client.h"
#include "deephaven/dhcore/chunk/chunk_maker.h"
#include "deephaven/dhcore/utility/utility.h"

using deephaven::client::Client;
using deephaven::client::TableHandle;
using deephaven::client::utility::TableMaker;
using deephaven::dhcore::DateTime;
using deephaven::dhcore::chunk::ChunkMaker;
using deephaven::dhcore::chunk::BooleanChunk;
using deephaven::dhcore::chunk::DateTimeChunk;
using deephaven::dhcore::chunk::DoubleChunk;
using deephaven::dhcore::chunk::FloatChunk;
using deephaven::dhcore::chunk::Int8Chunk;
using deephaven::dhcore::chunk::Int16Chunk;
using deephaven::dhcore::chunk::Int32Chunk;
using deephaven::dhcore::chunk::Int64Chunk;
using deephaven::dhcore::chunk::StringChunk;
using deephaven::dhcore::clienttable::ClientTable;
using deephaven::dhcore::container::RowSequence;
using deephaven::dhcore::utility::MakeReservedVector;

namespace deephaven::client::tests {
class CommonBase : public deephaven::dhcore::ticking::TickingCallback {
public:
  void OnFailure(std::exception_ptr ep) final {
    std::unique_lock guard(mutex_);
    exception_ptr_ = std::move(ep);
    cond_var_.notify_all();
  }

  std::pair<bool, std::exception_ptr> WaitForUpdate() {
    std::unique_lock guard(mutex_);
    while (true) {
      if (done_ || exception_ptr_ != nullptr) {
        return std::make_pair(done_, exception_ptr_);
      }
      cond_var_.wait(guard);
    }
  }

protected:
  void NotifyDone() {
    std::unique_lock guard(mutex_);
    done_ = true;
    cond_var_.notify_all();
  }

  std::mutex mutex_;
  std::condition_variable cond_var_;
  bool done_ = false;
  std::exception_ptr exception_ptr_;
};

class ReachesNRowsCallback final : public CommonBase {
public:
  explicit ReachesNRowsCallback(size_t target_rows) : target_rows_(target_rows) {}

  void OnTick(deephaven::dhcore::ticking::TickingUpdate update) final {
    std::cout << "=== The Full Table ===\n"
              << update.Current()->Stream(true, true)
              << '\n';
    if (update.Current()->NumRows() >= target_rows_) {
      NotifyDone();
    }
  }

private:
  size_t target_rows_ = 0;
};

TEST_CASE("Ticking Table eventually reaches 10 rows", "[ticking]") {
  const auto max_rows = 10;
  auto client = TableMakerForTests::CreateClient();
  auto tm = client.GetManager();

  auto table = tm.TimeTable(std::chrono::milliseconds(500)).Update("II = ii");
  table.BindToVariable("ticking");
  auto callback = std::make_shared<ReachesNRowsCallback>(max_rows);
  auto cookie = table.Subscribe(callback);

  while (true) {
    auto [done, eptr] = callback->WaitForUpdate();
    if (done) {
      break;
    }
    if (eptr != nullptr) {
      std::rethrow_exception(eptr);
    }
  }

  table.Unsubscribe(std::move(cookie));
}

class AllValuesGreaterThanNCallback final : public CommonBase {
public:
  explicit AllValuesGreaterThanNCallback(int64_t target) : target_(target) {}

  void OnTick(deephaven::dhcore::ticking::TickingUpdate update) final {
    const auto &current = update.Current();
    std::cout << "=== The Full Table ===\n"
              << current->Stream(true, true)
              << '\n';
    if (current->NumRows() == 0) {
      return;
    }
    auto col = current->GetColumn("Value", true);
    auto rs = current->GetRowSequence();
    auto data = Int64Chunk::Create(rs->Size());
    col->FillChunk(*rs, &data, nullptr);
    auto all_greater = true;
    for (auto element : data) {
      if (element < target_) {
        all_greater = false;
        break;
      }
    }
    if (all_greater) {
      NotifyDone();
    }
  }

private:
  int64_t target_ = 0;
};

TEST_CASE("Ticking Table modified rows are eventually all greater than 10", "[ticking]") {
  const int64_t target = 10;
  auto client = TableMakerForTests::CreateClient();
  auto tm = client.GetManager();

  auto table = tm.TimeTable(std::chrono::milliseconds(500))
      .View({"Key = (long)(ii % 10)", "Value = ii"})
      .LastBy("Key");
  auto callback = std::make_shared<AllValuesGreaterThanNCallback>(target);
  auto cookie = table.Subscribe(callback);

  while (true) {
    auto [done, eptr] = callback->WaitForUpdate();
    if (done) {
      break;
    }
    if (eptr != nullptr) {
      std::rethrow_exception(eptr);
    }
  }

  table.Unsubscribe(std::move(cookie));
}

class WaitForPopulatedTableCallback final : public CommonBase {
public:
  explicit WaitForPopulatedTableCallback(int64_t target) : target_(target) {}

  void OnTick(deephaven::dhcore::ticking::TickingUpdate update) final {
    const auto &current = update.Current();
    std::cout << "=== The Full Table ===\n"
        << current->Stream(true, true)
        << '\n';

    if (current->NumRows() < static_cast<size_t>(target_)) {
      // table not yet fully populated.
      return;
    }

    if (current->NumRows() > static_cast<size_t>(target_)) {
      // table has more rows than expected.
      auto message = fmt::format("Expected table to have {} rows, got {}",
          target_, current->NumRows());
      throw std::runtime_error(message);
    }

    auto int8s = MakeReservedVector<int8_t>(target_);
    auto int16s = MakeReservedVector<int16_t>(target_);
    auto int32s = MakeReservedVector<int32_t>(target_);
    auto int64s = MakeReservedVector<int64_t>(target_);
    auto floats = MakeReservedVector<float>(target_);
    auto doubles = MakeReservedVector<double>(target_);
    auto bools = MakeReservedVector<bool>(target_);
    auto strings = MakeReservedVector<std::string>(target_);
    auto date_times = MakeReservedVector<DateTime>(target_);

    auto date_time_start = DateTime::Parse("2001-03-01T12:34:56Z");

    for (int64_t i = 0; i != target_; ++i) {
      int8s.push_back(i);
      int16s.push_back(i);
      int32s.push_back(i);
      int64s.push_back(i);
      floats.push_back(i);
      doubles.push_back(i);
      bools.push_back((i % 2) == 0);
      strings.push_back(fmt::format("hello {}", i));
      date_times.push_back(DateTime::FromNanos(date_time_start.Nanos() + i));
    }

    CompareColumn<Int8Chunk>(*current, "Bytes", int8s);
    CompareColumn<Int16Chunk>(*current, "Shorts", int16s);
    CompareColumn<Int32Chunk>(*current, "Ints", int32s);
    CompareColumn<Int64Chunk>(*current, "Longs", int64s);
    CompareColumn<FloatChunk>(*current, "Floats", floats);
    CompareColumn<DoubleChunk>(*current, "Doubles", doubles);
    CompareColumn<StringChunk>(*current, "Strings", strings);
    CompareColumn<BooleanChunk>(*current, "Bools", bools);
    CompareColumn<DateTimeChunk>(*current, "DateTimes", date_times);

    NotifyDone();
  }

  template<typename ChunkType, typename T>
  void CompareColumn(const ClientTable &table, std::string_view column_name,
      const std::vector<T> &expected) {
    static_assert(std::is_same_v<typename ChunkType::value_type, T>);
    if (expected.size() != table.NumRows()) {
      auto message = fmt::format("Expected 'expected' to have size {}, have {}",
          table.NumRows(), expected.size());
      throw std::runtime_error(DEEPHAVEN_LOCATION_STR(message));
    }
    auto cs = table.GetColumn(column_name, true);
    auto rs = RowSequence::CreateSequential(0, table.NumRows());
    auto chunk = ChunkType::Create(table.NumRows());
    cs->FillChunk(*rs, &chunk, nullptr);

    for (size_t row_num = 0; row_num != expected.size(); ++row_num) {
      const auto &expected_elt = expected[row_num];
      const auto &actual_elt = chunk.data()[row_num];
      if (expected_elt != actual_elt) {
        auto message = fmt::format(R"(In column "{}", row {}, expected={}, actual={})",
            column_name, row_num, expected_elt, actual_elt);
        throw std::runtime_error(DEEPHAVEN_LOCATION_STR(message));
      }
    }
  }

private:
  int64_t target_ = 0;
};

TEST_CASE("Ticking Table all the data is eventually present", "[ticking]") {
  const int64_t target = 10;
  auto client = TableMakerForTests::CreateClient();
  auto tm = client.GetManager();

  auto table = tm.TimeTable("PT0:00:0.5")
      .Update({"II = (int)((ii * 7) % 10)",
          "Bytes = (byte)II",
          "Chars = (char)(II + 'a')",
          "Shorts = (short)II",
          "Ints = (int)II",
          "Longs = (long)II",
          "Floats = (float)II",
          "Doubles = (double)II",
          "Strings = `hello ` + II",
          "Bools = (II % 2) == 0",
          "DateTimes = '2001-03-01T12:34:56Z' + II"
      })
      .LastBy("II")
      .Sort(SortPair::Ascending("II"));

  auto callback = std::make_shared<WaitForPopulatedTableCallback>(target);
  auto cookie = table.Subscribe(callback);

  while (true) {
    auto [done, eptr] = callback->WaitForUpdate();
    if (done) {
      break;
    }
    if (eptr != nullptr) {
      std::rethrow_exception(eptr);
    }
  }

  table.Unsubscribe(std::move(cookie));
}
}  // namespace deephaven::client::tests

File: csharp/client/DhClientTests/TickingTest.cs
using Deephaven.DeephavenClient;
using Microsoft.VisualBasic;
using System.Runtime.InteropServices;
using System;
using Xunit.Abstractions;
using static System.Reflection.Metadata.BlobBuilder;

namespace Deephaven.DhClientTests;

public class TickingTest {
  private readonly ITestOutputHelper _output;

  public TickingTest(ITestOutputHelper output) {
    _output = output;
  }

  [Fact]
  public void EventuallyReaches10Rows() {
    const Int64 maxRows = 10;
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var thm = ctx.Client.Manager;

    using var table = thm.TimeTable(TimeSpan.FromMilliseconds(500)).Update("II = ii");
    var callback = new ReachesNRowsCallback(_output, maxRows);
    using var cookie = table.Subscribe(callback);

    while (true) {
      var (done, errorText) = callback.WaitForUpdate();
      if (done) {
        break;
      }
      if (errorText != null) {
        throw new Exception(errorText);
      }
    }

    table.Unsubscribe(cookie);
  }

  [Fact]
  public void AFourthTest() {
    Assert.Equal(7, 7);
  }

  [Fact]
  public void AFifthTest() {
    Assert.Equal(7, 7);
  }

  [Fact]
  public void AllEventuallyGreaterThan10() {
    const Int64 maxRows = 10;
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var thm = ctx.Client.Manager;

    using var table = thm.TimeTable(TimeSpan.FromMilliseconds(500))
      .View("Key = (long)(ii % 10)", "Value = ii")
      .LastBy("Key");

    var callback = new AllValuesGreaterThanNCallback(_output, maxRows);
    using var cookie = table.Subscribe(callback);

    while (true) {
      var (done, errorText) = callback.WaitForUpdate();
      if (done) {
        break;
      }
      if (errorText != null) {
        throw new Exception(errorText);
      }
    }

    table.Unsubscribe(cookie);
  }

  [Fact]
  public void AllDataEventuallyPresent() {
    const Int64 maxRows = 10;
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var thm = ctx.Client.Manager;

    using var table = thm.TimeTable("PT0:00:0.5")
      .Update(
        "II = (int)((ii * 7) % 10)",
        "Chars = II == 5 ? null : (char)(II + 'a')",
        "Bytes = II == 5 ? null : (byte)II",
        "Shorts = II == 5 ? null : (short)II",
        "Ints = II == 5 ? null : (int)II",
        "Longs = II == 5 ? null : (long)II",
        "Floats = II == 5 ? null : (float)II",
        "Doubles = II == 5 ? null : (double)II",
        "Bools = II == 5 ? null : ((II % 2) == 0)",
        "Strings = II == 5 ? null : (`hello ` + II)",
        "DateTimes = II == 5 ? null : ('2001-03-01T12:34:56Z' + II)")
      .LastBy("II")
      .Sort(SortPair.Ascending("II"))
      .DropColumns("Timestamp", "II");

    var callback = new WaitForPopulatedTableCallback(_output, maxRows);
    using var cookie = table.Subscribe(callback);

    while (true) {
      var (done, errorText) = callback.WaitForUpdate();
      if (done) {
        break;
      }
      if (errorText != null) {
        throw new Exception(errorText);
      }
    }

    table.Unsubscribe(cookie);
  }
}

public abstract class CommonBase : ITickingCallback {
  private readonly object _sync = new();
  private bool _done = false;
  private string? _errorText = null;

  public void OnFailure(string errorText) {
    lock (_sync) {
      _errorText = errorText;
      Monitor.PulseAll(_sync);
    }
  }

  public (bool, string?) WaitForUpdate() {
    lock (_sync) {
      while (true) {
        if (_done || _errorText != null) {
          return (_done, _errorText);
        }

        Monitor.Wait(_sync);
      }
    }
  }

  public abstract void OnTick(TickingUpdate update);

  protected void NotifyDone() {
    lock (_sync) {
      _done = true;
      Monitor.PulseAll(_sync);
    }
  }
}

public sealed class ReachesNRowsCallback : CommonBase {
  private readonly ITestOutputHelper _output;
  private readonly Int64 _targetRows;

  public ReachesNRowsCallback(ITestOutputHelper output, Int64 targetRows) {
    _output = output;
    _targetRows = targetRows;
  }

  public override void OnTick(TickingUpdate update) {
    _output.WriteLine($"=== The Full Table ===\n{update.Current.ToString(true, true)}");
    if (update.Current.NumRows >= _targetRows) {
      NotifyDone();
    }
  }
}

public sealed class WaitForPopulatedTableCallback : CommonBase {
  private readonly ITestOutputHelper _output;
  private readonly Int64 _target;

  public WaitForPopulatedTableCallback(ITestOutputHelper output, Int64 target) {
    _output = output;
    _target = target;
  }

  public override void OnTick(TickingUpdate update) {
    _output.WriteLine($"=== The Full Table ===\n{update.Current.ToString(true, true)}");

    var current = update.Current;

    if (current.NumRows < _target) {
      return;
    }

    if (current.NumRows > _target) {
      // table has more rows than expected.
      var message = $"Expected table to have {_target} rows, got {current.NumRows}";
      throw new Exception(message);
    }

    var charData = new List<char?>();
    var int8Data = new List<sbyte?>();
    var int16Data = new List<short?>();
    var int32Data = new List<int?>();
    var int64Data = new List<long?>();
    var floatData = new List<float?>();
    var doubleData = new List<double?>();
    var boolData = new List<bool?>();
    var stringData = new List<string?>();
    var dateTimeData = new List<DhDateTime?>();

    var dateTimeStart = DhDateTime.Parse("2001-03-01T12:34:56Z");

    for (Int64 i = 0; i != _target; ++i) {
      charData.Add((char)('a' + i));
      int8Data.Add((sbyte)(i));
      int16Data.Add((Int16)(i));
      int32Data.Add((Int32)(i));
      int64Data.Add((Int64)(i));
      floatData.Add((float)(i));
      doubleData.Add((double)(i));
      boolData.Add((i % 2) == 0);
      stringData.Add($"hello {i}");
      dateTimeData.Add(DhDateTime.FromNanos(dateTimeStart.Nanos + i));
    }

    if (_target == 0) {
      throw new Exception("Target should not be 0");
    }

    var t2 = (int)(_target / 2);
    // Set the middle element to the unset optional, which for the purposes of this test is
    // our representation of null.
    charData[t2] = null;
    int8Data[t2] = null;
    int16Data[t2] = null;
    int32Data[t2] = null;
    int64Data[t2] = null;
    floatData[t2] = null;
    doubleData[t2] = null;
    boolData[t2] = null;
    stringData[t2] = null;
    dateTimeData[t2] = null;

    var tc = new TableComparer();
    tc.AddColumn("Chars", charData);
    tc.AddColumn("Bytes", int8Data);
    tc.AddColumn("Shorts", int16Data);
    tc.AddColumn("Ints", int32Data);
    tc.AddColumn("Longs", int64Data);
    tc.AddColumn("Floats", floatData);
    tc.AddColumn("Doubles", doubleData);
    tc.AddColumn("Bools", boolData);
    tc.AddColumn("Strings", stringData);
    tc.AddColumn("DateTimes", dateTimeData);

    tc.AssertEqualTo(current);
    NotifyDone();
  }
}

public sealed class AllValuesGreaterThanNCallback : CommonBase {
  private readonly ITestOutputHelper _output;
  private readonly Int64 _target;

  public AllValuesGreaterThanNCallback(ITestOutputHelper output, Int64 target) {
    _output = output;
    _target = target;
  }

  public override void OnTick(TickingUpdate update) {
    _output.WriteLine($"=== The Full Table ===\n{update.Current.ToString(true, true)}");

    var current = update.Current;

    if (current.NumRows == 0) {
      return;
    }

    var (values, nulls) = current.GetColumn("Value");
    var data = (Int64[])values;
    var allGreater = data.All(elt => elt > _target);
    if (allGreater) {
      NotifyDone();
    }
  }
}
